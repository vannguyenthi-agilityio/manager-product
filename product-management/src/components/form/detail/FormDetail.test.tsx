// Libraries
import { fireEvent, render, screen } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import { MemoryRouter } from 'react-router-dom';

// Components
import FormDetail from '.';

const onBack = jest.fn();
const onSubmit = jest.fn();
const mockedUsedNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedUsedNavigate
}));

const setup = () => {
  const props = {
    handleSubmit: onSubmit,
    handleBack: onBack
  };

  return render(<FormDetail {...props} />, { wrapper: MemoryRouter });
};

describe('Product Form Component', () => {
  test('should render correctly', () => {
    setup();

    const formDetail = screen.getByTestId('form-detail');

    expect(formDetail).toMatchSnapshot();
  });

  test('should call onClick navigation to Home page', async () => {
    setup();

    const navHomePage = screen.getByTestId('click-to-home');
    expect(navHomePage).toBeDefined();

    fireEvent.click(navHomePage);

    expect(mockedUsedNavigate).toHaveBeenCalledTimes(1);
  });

  test('should call onClick pass validate to save Product ', async () => {
    jest.useFakeTimers();
    setup();

    const saveProduct = screen.getByTestId('click-save');

    expect(saveProduct).toBeDefined();

    const inputName = screen.getByTestId<HTMLInputElement>('input-name');
    expect(inputName).toBeDefined();

    const inputBrand = screen.getByTestId<HTMLInputElement>('input-brand');
    expect(inputBrand).toBeDefined();

    await act(async () => {
      fireEvent.change(inputName, { target: { value: 'Louis Vuitton' } });
      jest.runOnlyPendingTimers();
    });

    await act(async () => {
      fireEvent.change(inputBrand, { target: { value: 'Evan Flores' } });
      jest.runOnlyPendingTimers();
    });

    await act(async () => {
      fireEvent.click(saveProduct);
      jest.runOnlyPendingTimers();
    });

    expect(inputName.value).toBe('Louis Vuitton');
    expect(inputBrand.value).toBe('Evan Flores');
  });

  test('should call onClick do not pass validate show error message', async () => {
    jest.useFakeTimers();
    setup();

    const saveProduct = screen.getByTestId('click-save');
    expect(saveProduct).toBeDefined();

    const inputName = screen.getByTestId<HTMLInputElement>('input-name');
    expect(inputName).toBeDefined();

    const inputBrand = screen.getByTestId<HTMLInputElement>('input-brand');
    expect(inputBrand).toBeDefined();

    const inputQuantity = screen.getByTestId<HTMLInputElement>('input-quantity');
    expect(inputQuantity).toBeDefined();
    expect(saveProduct).toBeDisabled;

    const inputPrice = screen.getByTestId<HTMLInputElement>('input-price');
    expect(inputPrice).toBeDefined();

    await act(async () => {
      fireEvent.change(inputName, { target: { value: '' } });
      jest.runOnlyPendingTimers();
    });

    await act(async () => {
      fireEvent.change(inputBrand, { target: { value: '' } });
      jest.runOnlyPendingTimers();
    });

    await act(async () => {
      fireEvent.change(inputQuantity, { target: { value: '' } });
      jest.runOnlyPendingTimers();
    });

    await act(async () => {
      fireEvent.change(inputPrice, { target: { value: '' } });
      jest.runOnlyPendingTimers();
    });

    await act(async () => {
      fireEvent.click(saveProduct);
      jest.runOnlyPendingTimers();
    });
    expect(onSubmit).toHaveBeenCalledTimes(0);

    expect(inputName.value).toBe('');
    expect(inputBrand.value).toBe('');
  });
});
