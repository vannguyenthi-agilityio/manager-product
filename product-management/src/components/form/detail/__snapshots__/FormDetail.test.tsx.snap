// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Form Component should render correctly 1`] = `
<div
  class="css-0"
  data-testid="form-detail"
>
  <div
    class="css-6jf9kx"
    data-testid="form-product"
  >
    <div
      class="chakra-form-control css-3qdq65"
      role="group"
    >
      <label
        class="chakra-form__label css-1gtigks"
        for="name"
        id="field-:r0:-label"
      >
        Name
      </label>
      <input
        class="chakra-input css-0"
        data-testid="input-name"
        id="name"
        name="product.name"
        placeholder="Enter Name..."
      />
    </div>
    <div
      class="chakra-form-control css-3qdq65"
      role="group"
    >
      <label
        class="chakra-form__label css-1gtigks"
        for="quantity"
        id="field-:r1:-label"
      >
        Quantity
      </label>
      <input
        class="chakra-input css-0"
        data-testid="input-quantity"
        id="quantity"
        name="product.quantity"
        placeholder="Quantity"
      />
    </div>
    <div
      class="chakra-form-control css-3qdq65"
      role="group"
    >
      <label
        class="chakra-form__label css-1gtigks"
        for="price"
        id="field-:r2:-label"
      >
        Price
      </label>
      <input
        class="chakra-input css-0"
        data-testid="input-price"
        id="price"
        name="product.price"
        placeholder="Price"
      />
    </div>
    <div
      class="chakra-stack css-1y87ypt"
    >
      <div
        class="chakra-form-control css-1ryeq9i"
        role="group"
      >
        <label
          class="chakra-form__label css-1gtigks"
          for="status"
          id="field-:r3:-label"
        >
          Status
        </label>
        <label>
          <div
            class="chakra-select__wrapper css-42b2qy"
          >
            <select
              class="chakra-select css-1ik61og"
              data-testid="select"
              id="status"
              name="product.status"
            >
              <option
                selected=""
                value="available"
              >
                Available
              </option>
              <option
                value="sold_out"
              >
                Sold out
              </option>
            </select>
            <div
              class="chakra-select__icon-wrapper css-162mkon"
            >
              <svg
                aria-hidden="true"
                class="chakra-select__icon"
                focusable="false"
                role="presentation"
                style="width: 1em; height: 1em; color: currentColor;"
                viewBox="0 0 24 24"
              >
                <path
                  d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                  fill="currentColor"
                />
              </svg>
            </div>
          </div>
        </label>
      </div>
      <div
        class="chakra-form-control css-0"
        role="group"
        style="max-width: 200px;"
      >
        <label
          class="chakra-form__label css-1gtigks"
          for="status"
          id="field-:r4:-label"
        >
          Types
        </label>
        <label>
          <div
            class="chakra-select__wrapper css-42b2qy"
          >
            <select
              class="chakra-select css-1ik61og"
              data-testid="select"
              id="type"
              name="product.types"
            >
              <option
                selected=""
                value="Gemstone"
              >
                Gemstone
              </option>
              <option
                value="metal"
              >
                Metal
              </option>
              <option
                value="diamond"
              >
                Diamond
              </option>
              <option
                value="gold"
              >
                Gold
              </option>
              <option
                value="ring"
              >
                Ring
              </option>
            </select>
            <div
              class="chakra-select__icon-wrapper css-162mkon"
            >
              <svg
                aria-hidden="true"
                class="chakra-select__icon"
                focusable="false"
                role="presentation"
                style="width: 1em; height: 1em; color: currentColor;"
                viewBox="0 0 24 24"
              >
                <path
                  d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                  fill="currentColor"
                />
              </svg>
            </div>
          </div>
        </label>
      </div>
    </div>
    <div
      class="chakra-form-control css-3qdq65"
      role="group"
    >
      <label
        class="chakra-form__label css-1gtigks"
        for="name"
        id="field-:r5:-label"
      >
        Brand
      </label>
      <input
        class="chakra-input css-0"
        data-testid="input-brand"
        id="brand"
        name="product.brand"
        placeholder="Enter Brand..."
      />
    </div>
  </div>
  <div
    class="chakra-stack css-1y87ypt"
  >
    <button
      class="chakra-button css-168qeg6"
      data-testid="click-to-home"
      type="button"
    >
      Back
    </button>
    <button
      class="chakra-button css-1jitvwb"
      data-testid="click-save"
      disabled=""
      style="cursor: not-allowed;"
      type="button"
    >
      Save
    </button>
  </div>
</div>
`;
